### PR TITLE
Optimization changes to completely fold OptionSet literals

### PR DESCRIPTION
**Pull request by eeckstein**
With this the optimizer is able to fold OptionSet literals like
     [.first, .second, .third, .fourth]
to a single integer constant. See test/SILOptimizer/optionset.swift

This PR consists of a couple of commits, which are (mostly) small changes to existing optimizations:

* DeadObjectElimination: handle objects for which the destructor is de-virtualized and inlined
* Tuning the inliner heuristics
* ARCCodeMotion: add a small peephole optimization which removes adjacent retain/release pairs.
* GlobalOpt: Move the object outlining from GlobalOpt into a separate pass "ObjectOutliner"
* SideEffectAnalysis: don't assume the worst side-effects for a release instruction
* SideEffectAnalysis: handle unchecked_ref_cast instructions
* SimplifyCFG: do constant folding also in SimplifyCFG.
* Move the constant folding logic into a separate utility which can be used by multiple passes.

For details see the commit messages

rdar://problem/26411312
and also rdar://problem/34758773

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci smoke benchmark

**Comment by eeckstein** [comment]
@swift-ci Please test compiler performance

File: lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp [hunk 1]
```diff
@@ -347,10 +347,6 @@ void SideEffectAnalysis::analyzeInstruction(FunctionInfo *FInfo,
     case SILInstructionKind::ReleaseValueInst:
     case SILInstructionKind::UnownedReleaseInst:
       FInfo->FE.getEffectsOn(I->getOperand(0))->Releases = true;
-      
-      // TODO: Check the call graph to be less conservative about what
-      // destructors might be called.
-      FInfo->FE.setWorstEffects();
```
Comment (jrose-apple) [inline lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp]: I don't see how this is safe in general. Shouldn't the client have to say explicitly when they know a value hasn't escaped?

**Comment by eeckstein** [comment]
@swift-ci Please test compiler performance

File: lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp]: FunctionEffects::getMemBehavior() returns the worst effects anyway if one of the parameter effects contains a "release".
The only thing which changed is that if a client explicitly looks at parameter effects, the client has to consider that a "released" parameter also potentially means worst effects on other parameters.

Comment (jrose-apple) [reply lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp]: Hm. It feels like it's too easy for someone to forget that, but I'll have to trust that the design makes sense for SideEffectAnalysis.

Comment (eeckstein) [reply lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp]: Well, I should add a comment to the ABI functions

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci Please smoke test compiler performance

**Comment by eeckstein** [comment]
@swift-ci Please smoke test compiler performance

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci Please smoke test compiler performance

**Comment by eeckstein** [comment]
@swift-ci Please smoke test compiler performance

**Comment by shahmishal** [comment]
@swift-ci Please smoke test compiler performance

**Comment by shahmishal** [comment]
@swift-ci Please smoke test compiler performance

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Pull request merged**